lib/db2.js

```diff
@@ -442,80 +443,66 @@ DB2.prototype.update = function(model, where, data, options, cb) {
 DB2.prototype.updateOrCreate = DB2.prototype.save =
   function(model, data, options, callback) {
     var self = this;
-    var fields = self.buildFields(model, data);
     var id = self.idName(model);
-    var sql = new ParameterizedSQL('MERGE INTO ' +
-                                   self.schema.toUpperCase() + '.' +
-                                   self.tableEscaped(model));
-    var columnValues = fields.columnValues;
-    var fieldNames = fields.names;
-    var setValues = [];
-
-    var definition = self.getModelDefinition(model);
-
-    if (fieldNames.length) {
-      sql.merge('AS MT(');
-
-      Object.keys(definition.properties).forEach(function(prop) {
-        setValues.push(new ParameterizedSQL(self.columnEscaped(model, prop)));
-      });
-
-      var columns = ParameterizedSQL.join(setValues, ',');
-      sql.merge(columns);
-      sql.merge(')');
-      sql.merge('USING (SELECT * FROM TABLE( VALUES(');
-
-      setValues = [];
-      for (var i = 0, n = fields.names.length; i < n; i++) {
-        setValues.push(new ParameterizedSQL('CAST (' + columnValues[i].sql +
-          ' AS ' +
-          self.buildColumnType(fields.properties[i]) + ')',
-          columnValues[i].params));
-
-        if (i < (n - 1))
-          setValues[i].sql = setValues[i].sql + ',';
-      }
-
-      sql.merge(setValues);
-      sql.merge('))) AS VT(' + fieldNames.join(',') + ')');
-      sql.merge('ON');
-      sql.merge('(MT.\"' + id + '\" = VT.\"' + id + '\")');
-      sql.merge('WHEN NOT MATCHED THEN INSERT (' +
-      fieldNames.join(',') + ')');
-
-      var values = ParameterizedSQL.join(columnValues, ',');
-      values.sql = 'VALUES(' + values.sql + ')';
-      sql.merge(values);
-    } else {
-      sql.merge(self.buildInsertDefaultValues(model, data, options));
-    }
-
-    sql.merge('WHEN MATCHED THEN UPDATE SET');
-
-    setValues = [];
-    for (i = 0, n = fields.names.length; i < n; i++) {
-      if (!fields.properties[i].id) {
-        setValues.push(new ParameterizedSQL(fields.names[i] + '=' +
-          '(' + columnValues[i].sql + ')', columnValues[i].params));
+    var stmt;
+    var tableName = self.tableEscaped(model);
+    var idData = {};
+    var countData;
+    var isInsert = false;
+    var meta = {};
+    var connection;
+    var currentTx;
+
+    process.nextTick(function() {
+      if (options.transaction) {
+        connection = options.transaction.connection;
+      } else {
+        try {
+          connection = Driver.openSync(self.connStr);
```

Comment:
MySQL is never in the situation where a function such as updateOrCreate can be called either with or without an active transaction specified in options because the executed SQL is a single atomic operation.  To make this fully async we would need to be able to distinguish between being called from within an active transaction vs. needing to start a new transaction since the multiple calls to executeSQL must be atomic from the callers perspective.

I'll poke around more but at the moment I don't see how changing to async will be possible.